internal/cluster/connection_manager.go

```diff
@@ -0,0 +1,655 @@
+	bundle.Check()
+	// TODO: make circuit breaker configurable
+	circuitBreaker := cb.NewCircuitBreaker(
+		cb.MaxRetries(math.MaxInt32),
+		cb.MaxFailureCount(3),
+		cb.RetryPolicy(func(attempt int) time.Duration {
+			if attempt < 10 {
+				return time.Duration((attempt+1)*100) * time.Millisecond
+			}
+			if attempt < 1000 {
+				return time.Duration(attempt*attempt) * time.Millisecond
+			}
+			return 20 * time.Minute
+		}))
+	manager := &ConnectionManager{
+		requestCh:            bundle.RequestCh,
+		responseCh:           bundle.ResponseCh,
+		clusterService:       bundle.ClusterService,
+		partitionService:     bundle.PartitionService,
+		serializationService: bundle.SerializationService,
+		eventDispatcher:      bundle.EventDispatcher,
+		invocationFactory:    bundle.InvocationFactory,
+		clusterConfig:        bundle.ClusterConfig,
+		credentials:          bundle.Credentials,
+		clientName:           bundle.ClientName,
+		clientUUID:           internal.NewUUID(),
+		connMap:              newConnectionMap(),
+		addressTranslator:    NewDefaultAddressTranslator(),
+		smartRouting:         bundle.SmartRouting,
+		logger:               bundle.Logger,
+		doneCh:               make(chan struct{}, 1),
+		cb:                   circuitBreaker,
+	}
+	return manager
+}
+
+func (m *ConnectionManager) Start() error {
+	if !atomic.CompareAndSwapInt32(&m.state, created, starting) {
+		return nil
+	}
+	if _, err := m.cb.Try(func(ctx context.Context) (interface{}, error) {
+		return nil, m.connectCluster()
+	}).Result(); err != nil {
+		return err
+	}
+	connectionClosedSubscriptionID := event.MakeSubscriptionID(m.handleConnectionClosed)
+	m.eventDispatcher.Subscribe(EventConnectionClosed, connectionClosedSubscriptionID, m.handleConnectionClosed)
+	memberAddedSubscriptionID := event.MakeSubscriptionID(m.handleMembersAdded)
+	m.eventDispatcher.Subscribe(EventMembersAdded, memberAddedSubscriptionID, m.handleMembersAdded)
+	go m.doctor()
+	if m.logger.CanLogDebug() {
+		go m.logStatus()
+	}
+	atomic.StoreInt32(&m.state, ready)
+	m.eventDispatcher.Publish(NewConnected())
+	return nil
+}
+
+func (m *ConnectionManager) Stop() chan struct{} {
+	if !atomic.CompareAndSwapInt32(&m.state, ready, stopping) {
+		close(m.doneCh)
+		return m.doneCh
+	}
+	go func() {
+		connectionClosedSubscriptionID := event.MakeSubscriptionID(m.handleConnectionClosed)
+		m.eventDispatcher.Unsubscribe(EventConnectionClosed, connectionClosedSubscriptionID)
+		memberAddedSubscriptionID := event.MakeSubscriptionID(m.handleMembersAdded)
+		m.eventDispatcher.Unsubscribe(EventMembersAdded, memberAddedSubscriptionID)
+		m.connMap.CloseAll()
+		atomic.StoreInt32(&m.state, stopped)
+		close(m.doneCh)
+	}()
+	return m.doneCh
+}
+
+func (m *ConnectionManager) NextConnectionID() int64 {
+	return atomic.AddInt64(&m.nextConnID, 1)
+}
+
+func (m *ConnectionManager) GetConnectionForAddress(addr *pubcluster.AddressImpl) *Connection {
+	return m.connMap.GetConnectionForAddr(addr.String())
```

Comment:
nit: both `AddressImpl.String()` and `UUID.String()` used in this and other methods of the connection manager, generate litter on each operation. That could be mitigated by caching the string in the struct and returning it from the `String()` method.

Reply:
This seems like something Go's GC can easily handle. I think the biggest benefit will be in the client message side as you pointed out.

Reply:
GC can handle a lot of litter structs for sure. But users will appreciate less stress put on the GC by the library. We're talking of low-hanging fruits here and, since Hazelcast client libraries claim to provide high performance, such optimizations should be included. At least, that's how I see.